## 生成器(Generator)

1. 什么是生成器？

生成器是一个通过构造函数Generator创建的函数，生成器既是一个迭代器，同时又是一个可迭代对象

2. 如何创建一个生成器

生成器的创建必须使用生成器创建函数(Generator Function)

3. 如何书写一个生成器函数？
```js
//这是一个生成器函数 该函数一定返回一个生成器
function *method(){

}
```
4. 生成器函数内部是如何执行的？
生成器函数内部是为了给生成器的每次迭代提供数据的，当调用函数得到一个生成器的时候，里面的代码不会运行，只有调用next方法 才会执行里面的代码 到 yield关键字的位置
```js
function *methods(){
    console.log('第一次运行')
    yield 1
    console.log('第二次运行')
    yield 2
    console.log('第三次运行')
}
const generator = methods()  //啥都不输出 因为不会运行里面的代码
console.log(generator.next()) // 第一次运行 {value : 1 , done :false}
console.log(generator.next()) // 第二次运行 {value : 2 , done :false}
console.log(generator.next()) // 第三次运行 {value : undefined , done :true}
```

生成器书写斐波那契数列
```js
function* iteratorFBNX() {
    let pre1 = 1;
    let pre2 = 1;
    let n = 1
    while(true){
        if(n <= 2 ){
            yield 1
        }else{
            const newValue = pre1 + pre2
            yield newValue
            pre2 = pre1
            pre1 = newValue 
        }
        n ++
    }
}
var itra1 = iteratorFBNX()
```

细节
- 生成器可以有返回值
  - 生成器的返回值出现在**第一次**done为true的value位置中
```js
function *methods(){
    console.log('第一次运行')
    yield 1
    console.log('第二次运行')
    yield 2
    console.log('第三次运行')
    return 10
}
const generator = methods()  //啥都不输出 因为不会运行里面的代码
console.log(generator.next()) // 第一次运行 {value : 1 , done :false}
console.log(generator.next()) // 第二次运行 {value : 2 , done :false}
console.log(generator.next()) // 第三次运行 {value : 10 , done :true} 返回值第一次有效
console.log(generator.next()) // 第三次运行 {value : undefined , done :true}
```
- 调用next方法时 可以传递参数 传递的参数会交给yield表达式的返回值  注意:**第一次调用next方法时，传参没有意义**
- 在生成器函数内部可以调用其它生成器函数 但是必须加上*
```js
function* test2() {
    yield 1
    yield 2
}
function* test1() {
    yield* test2()
    yield 3
    yield 4
}
const generator = test1()
```



## 生成器的其他API
- return 可以通过调用该方法 提前结束生成器函数 从而提前让整个迭代过程结束
  - return中还可以传递参数 就是结束时的value值
```js
function *methods(){
    yield 1
    yield 2
}
const generator = methods()
generator.next()  // {value : 1, done :true}
generator.return()  // {value : undefined , done :true}
```
- throw 方法 调用该方法可以在生成器中产生一个错误

